README.md
<img src="doc/asset/Rx_Logo_S.png" alt="RxJS Logo" width="86" height="86"> RxJS: Reactive Extensions For JavaScript
======================================


[![CircleCI](https://circleci.com/gh/ReactiveX/rxjs/tree/master.svg?style=svg)](https://circleci.com/gh/ReactiveX/rxjs/tree/master)
[![npm version](https://badge.fury.io/js/rxjs.svg)](http://badge.fury.io/js/rxjs)
[![Join the chat at https://gitter.im/Reactive-Extensions/RxJS](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/Reactive-Extensions/RxJS?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

# RxJS 7 (alpha)

### FOR 6.X PLEASE GO TO [THE 6.x BRANCH](https://github.com/zlq4863947/rxjs-cn//tree/6.x)

Reactive Extensions Library for JavaScript. This is a rewrite of [Reactive-Extensions/RxJS](https://github.com/Reactive-Extensions/RxJS) and is the latest production-ready version of RxJS. This rewrite is meant to have better performance, better modularity, better debuggable call stacks, while staying mostly backwards compatible, with some breaking changes that reduce the API surface.

[Apache 2.0 License](LICENSE.txt)

- [Code of Conduct](CODE_OF_CONDUCT.md)
- [Contribution Guidelines](CONTRIBUTING.md)
- [Maintainer Guidelines](doc/maintainer-guidelines.md)
- [Creating Operators](doc/operator-creation.md)
- [API Documentation (WIP)](https://rxjs.dev/)

## Versions In This Repository

- [master](https://github.com/zlq4863947/rxjs-cn//commits/master) - This is all of the current, unreleased work, which is against v7 of RxJS right now
- [stable](https://github.com/zlq4863947/rxjs-cn//tree/6.x) - This is the branch for the latest version you'd get if you do `npm install rxjs`

Most PRs should be made to **master**.

## Important

By contributing or commenting on issues in this repository, whether you've read them or not, you're agreeing to the [Contributor Code of Conduct](CODE_OF_CONDUCT.md). Much like traffic laws, ignorance doesn't grant you immunity.

## Installation and Usage

### ES6 via npm

```sh
npm install rxjs
```

It's recommended to pull in the Observable creation methods you need directly from `'rxjs'` as shown below with `range`. And you can pull in any operator you need from one spot, under `'rxjs/operators'`.

```ts
import { range } from 'rxjs';
import { map, filter } from 'rxjs/operators';

range(1, 200).pipe(
  filter(x => x % 2 === 1),
  map(x => x + x)
).subscribe(x => console.log(x));
```

### CDN

For CDN, you can use [unpkg](https://unpkg.com/):

https://unpkg.com/rxjs/bundles/rxjs.umd.min.js

The global namespace for rxjs is `rxjs`:

```js
const { range } = rxjs;
const { map, filter } = rxjs.operators;

range(1, 200).pipe(
  filter(x => x % 2 === 1),
  map(x => x + x)
).subscribe(x => console.log(x));
```

## Goals

- Smaller overall bundles sizes
- Provide better performance than preceding versions of RxJS
- To model/follow the [Observable Spec Proposal](https://github.com/zenparsing/es-observable) to the observable
- Provide more modular file structure in a variety of formats
- Provide more debuggable call stacks than preceding versions of RxJS

## Building/Testing

- `npm run build_all` - builds everything
- `npm test` - runs tests
- `npm run test_no_cache` - run test with `ts-node` set to false

## Performance Tests

Run `npm run build_perf` or `npm run perf` to run the performance tests with `protractor`.

Run `npm run perf_micro [operator]` to run micro performance test benchmarking operator.

## Adding documentation
We appreciate all contributions to the documentation of any type. All of the information needed to get the docs app up and running locally as well as how to contribute can be found in the [documentation directory](./docs_app).


CHANGELOG.md
# [7.0.0-alpha.1](https://github.com/zlq4863947/rxjs-cn//compare/7.0.0-alpha.0...7.0.0-alpha.1) (2019-12-27)


### Bug Fixes

* chain subscriptions from interop observables ([#5059](https://github.com/zlq4863947/rxjs-cn//issues/5059)) ([d7f7078](https://github.com/zlq4863947/rxjs-cn//commit/d7f7078))
* clear subscription on shareReplay completion ([#5044](https://github.com/zlq4863947/rxjs-cn//issues/5044)) ([35e600f](https://github.com/zlq4863947/rxjs-cn//commit/35e600f)), closes [#5034](https://github.com/zlq4863947/rxjs-cn//issues/5034)
* **closure:** Annotate next() for ReplaySubject ([#5088](https://github.com/zlq4863947/rxjs-cn//issues/5088)) ([8687fbd](https://github.com/zlq4863947/rxjs-cn//commit/8687fbd))
* **closure:** static prop frameTimeFactor being collapsed when compiled with closure. ([39872c9](https://github.com/zlq4863947/rxjs-cn//commit/39872c9))
* **docs:** remove repetitive op3() in example ([#5043](https://github.com/zlq4863947/rxjs-cn//issues/5043)) ([e17df33](https://github.com/zlq4863947/rxjs-cn//commit/e17df33))
* **filter:** Fix overload order for filter to support inferring the generic type ([#5024](https://github.com/zlq4863947/rxjs-cn//issues/5024)) ([8255365](https://github.com/zlq4863947/rxjs-cn//commit/8255365))
* **fromFetch:** passing already aborted signal to init aborts fetch ([0e4849a](https://github.com/zlq4863947/rxjs-cn//commit/0e4849a))


### Features

* **concatWith:** adds concatWith ([#4988](https://github.com/zlq4863947/rxjs-cn//issues/4988)) ([dc89736](https://github.com/zlq4863947/rxjs-cn//commit/dc89736))



# [7.0.0-alpha.0](https://github.com/zlq4863947/rxjs-cn//compare/6.5.2...7.0.0-alpha.0) (2019-09-18)


### Bug Fixes

* missing package.json in rxjs/fetch ([#5001](https://github.com/zlq4863947/rxjs-cn//issues/5001)) ([f4bee07](https://github.com/zlq4863947/rxjs-cn//commit/f4bee07))
* **filter:** Resolve TS build failures for certain situations where Boolean is the predicate ([77c7dfd](https://github.com/zlq4863947/rxjs-cn//commit/77c7dfd))
* **pluck:** key union type strictness ([#4585](https://github.com/zlq4863947/rxjs-cn//issues/4585)) ([bd5ec2d](https://github.com/zlq4863947/rxjs-cn//commit/bd5ec2d))
* **race:** ignore latter sources after first complete or error ([#4809](https://github.com/zlq4863947/rxjs-cn//issues/4809)) ([f31c3df](https://github.com/zlq4863947/rxjs-cn//commit/f31c3df)), closes [#4808](https://github.com/zlq4863947/rxjs-cn//issues/4808)
* **scan/reduce:** Typings correct for mixed seed/value types ([#4858](https://github.com/zlq4863947/rxjs-cn//issues/4858)) ([b89ebe5](https://github.com/zlq4863947/rxjs-cn//commit/b89ebe5))
* **scheduled:** import from relative paths ([#4832](https://github.com/zlq4863947/rxjs-cn//issues/4832)) ([1d37a87](https://github.com/zlq4863947/rxjs-cn//commit/1d37a87))
* **TS:** Error impls now properly type `this` ([#4978](https://github.com/zlq4863947/rxjs-cn//issues/4978)) ([7606dc7](https://github.com/zlq4863947/rxjs-cn//commit/7606dc7))
* **TS:** fix type inference for defaultIfEmpty. ([#4833](https://github.com/zlq4863947/rxjs-cn//issues/4833)) ([9b5ce2f](https://github.com/zlq4863947/rxjs-cn//commit/9b5ce2f))
* **types:** add Boolean signature to filter ([#4961](https://github.com/zlq4863947/rxjs-cn//issues/4961)) ([259853e](https://github.com/zlq4863947/rxjs-cn//commit/259853e)), closes [#4959](https://github.com/zlq4863947/rxjs-cn//issues/4959) [/github.com/ReactiveX/rxjs/issues/4959#issuecomment-520629091](https://github.com//github.com/ReactiveX/rxjs/issues/4959/issues/issuecomment-520629091)


### Features

* **animationFrames:** Adds an observable of animationFrames ([#5021](https://github.com/zlq4863947/rxjs-cn//issues/5021)) ([6a4cd68](https://github.com/zlq4863947/rxjs-cn//commit/6a4cd68))
* **concat:** can infer N types ([6c0cbc4](https://github.com/zlq4863947/rxjs-cn//commit/6c0cbc4))
* **of:** Update of typings ([e8adbb5](https://github.com/zlq4863947/rxjs-cn//commit/e8adbb5))
* **rxjs-compat:** removed for v7 ([#4839](https://github.com/zlq4863947/rxjs-cn//issues/4839)) ([79b1b95](https://github.com/zlq4863947/rxjs-cn//commit/79b1b95))
* **TestScheduler:** expose `frameTimeFactor` property ([#4977](https://github.com/zlq4863947/rxjs-cn//issues/4977)) ([8c32ed0](https://github.com/zlq4863947/rxjs-cn//commit/8c32ed0))
* **TS:** Update to TypeScript 3.5.3 ([741a136](https://github.com/zlq4863947/rxjs-cn//commit/741a136))


### BREAKING CHANGES

* **concat:** Generic signature changed. Recommend not explicitly passing generics, just let inference do its job. If you must, cast with `as`.
* **of:** Generic signature changed, do not specify generics, allow them to be inferred or use `as`
* **of:** Use with more than 9 arguments, where the last argument is a `SchedulerLike` may result in the wrong type which includes the `SchedulerLike`, even though the run time implementation does not support that. Developers should be using `scheduled` instead
* **TS:** RxJS requires TS 3.5
* **rxjs-compat:** `rxjs/Rx` is no longer a valid import site.
* **rxjs-compat:** `rxjs-compat` is not published for v7 (yet)
* **race:** `race()` will no longer subscribe to subsequent observables if a provided source synchronously errors or completes. This means side effects that might have occurred during subscription in those rare cases will no longer occur.



## [6.5.3](https://github.com/zlq4863947/rxjs-cn//compare/6.5.2...6.5.3) (2019-09-03)

### Bug Fixes

* **general:** Refactor modules so they don't show side effects in some tools ([#4769](https://github.com/zlq4863947/rxjs-cn//issues/4769)) ([9829c5e0](https://github.com/zlq4863947/rxjs-cn//commit/9829c5e0))
* **defer:** restrict allowed factory types ([#4835](https://github.com/zlq4863947/rxjs-cn//issues/4835)) ([40a22096](https://github.com/zlq4863947/rxjs-cn//commit/40a22096))

## [6.5.2](https://github.com/zlq4863947/rxjs-cn//compare/6.5.0...6.5.2) (2019-05-10)


### Bug Fixes


spec/operators/catch-spec.ts
import { expect } from 'chai';
import { concat, defer, Observable, of, throwError, EMPTY, from } from 'rxjs';
import { catchError, delay, map, mergeMap, takeWhile } from 'rxjs/operators';
import * as sinon from 'sinon';
import { createObservableInputs } from '../helpers/test-helper';
import { TestScheduler } from 'rxjs/testing';
import { observableMatcher } from '../helpers/observableMatcher';
import { asInteropObservable } from '../helpers/interop-helper';

declare function asDiagram(arg: string): Function;

/** @test {catch} */
describe('catchError operator', () => {
  let testScheduler: TestScheduler;

  beforeEach(() => {
    testScheduler = new TestScheduler(observableMatcher);
  });

  asDiagram('catch')('should catch error and replace with a cold Observable', () => {
    testScheduler.run(({ hot, cold, expectObservable }) => {
      const e1 = hot('  --a--b--#       ');
      const e2 = cold('         -1-2-3-|');
      const expected = '--a--b---1-2-3-|';

      const result = e1.pipe(catchError((err: any) => e2));

      expectObservable(result).toBe(expected);
    });
  });

  it('should catch error and replace it with Observable.of()', () => {
    testScheduler.run(({ hot, expectObservable, expectSubscriptions }) => {
      const e1 = hot('  --a--b--c--------|');
      const subs = '    ^-------!';
      const expected = '--a--b--(XYZ|)';

      const result = e1.pipe(
        map((n: string) => {
          if (n === 'c') {
            throw 'bad';
          }
          return n;
        }),
        catchError((err: any) => {
          return of('X', 'Y', 'Z');
        })
      );

      expectObservable(result).toBe(expected);
      expectSubscriptions(e1.subscriptions).toBe(subs);
    });
  });

  it('should catch error and replace it with a cold Observable', () => {
    testScheduler.run(({ hot, cold, expectObservable, expectSubscriptions }) => {
      const e1 = hot('  --a--b--#          ');
      const e1subs = '  ^-------!          ';
      const e2 = cold('         1-2-3-4-5-|');
      const e2subs = '  --------^---------!';
      const expected = '--a--b--1-2-3-4-5-|';

      const result = e1.pipe(catchError((err: any) => e2));

      expectObservable(result).toBe(expected);
      expectSubscriptions(e1.subscriptions).toBe(e1subs);
      expectSubscriptions(e2.subscriptions).toBe(e2subs);
    });
  });

  it('should allow unsubscribing explicitly and early', () => {
    testScheduler.run(({ hot, expectObservable, expectSubscriptions }) => {
      const e1 = hot('  --1-2-3-4-5-6---#');
      const e1subs = '  ^------!         ';
      const expected = '--1-2-3-         ';
      const unsub = '   -------!         ';

      const result = e1.pipe(
        catchError(() => {
          return of('X', 'Y', 'Z');
        })
      );

      expectObservable(result, unsub).toBe(expected);
      expectSubscriptions(e1.subscriptions).toBe(e1subs);
    });
  });

  it('should not break unsubscription chain when unsubscribed explicitly', () => {
    testScheduler.run(({ hot, expectObservable, expectSubscriptions }) => {
      const e1 = hot('  --1-2-3-4-5-6---#');
      const e1subs = '  ^------!         ';
      const expected = '--1-2-3-         ';
      const unsub = '   -------!         ';

      const result = e1.pipe(
        mergeMap((x: any) => of(x)),
        catchError(() => {
          return of('X', 'Y', 'Z');
        }),
        mergeMap((x: any) => of(x))
      );

      expectObservable(result, unsub).toBe(expected);
      expectSubscriptions(e1.subscriptions).toBe(e1subs);
    });
  });

  it('should unsubscribe from a caught hot caught observable when unsubscribed explicitly', () => {
    testScheduler.run(({ hot, expectObservable, expectSubscriptions }) => {
      const e1 = hot('  -1-2-3-#          ');
      const e1subs = '  ^------!          ';
      const e2 = hot('  ---3-4-5-6-7-8-9-|');
      const e2subs = '  -------^----!     ';
      const expected = '-1-2-3-5-6-7-     ';
      const unsub = '   ------------!     ';

      const result = e1.pipe(catchError(() => e2));

      expectObservable(result, unsub).toBe(expected);
      expectSubscriptions(e1.subscriptions).toBe(e1subs);
      expectSubscriptions(e2.subscriptions).toBe(e2subs);
    });
  });

  it('should unsubscribe from a caught cold caught observable when unsubscribed explicitly', () => {
    testScheduler.run(({ hot, cold, expectObservable, expectSubscriptions }) => {
      const e1 = hot('  -1-2-3-#          ');
      const e1subs = '  ^------!          ';
      const e2 =  cold('       5-6-7-8-9-|');
      const e2subs = '  -------^----!     ';
      const expected = '-1-2-3-5-6-7-     ';
      const unsub = '   ------------!     ';

      const result = e1.pipe(catchError(() => e2));

      expectObservable(result, unsub).toBe(expected);
      expectSubscriptions(e1.subscriptions).toBe(e1subs);
      expectSubscriptions(e2.subscriptions).toBe(e2subs);
    });
  });

  it('should unsubscribe from a caught cold caught interop observable when unsubscribed explicitly', () => {
    testScheduler.run(({ hot, cold, expectObservable, expectSubscriptions }) => {
      const e1 = hot('  -1-2-3-#          ');
      const e1subs = '  ^------!          ';
      const e2 =  cold('       5-6-7-8-9-|');
      const e2subs = '  -------^----!     ';
      const expected = '-1-2-3-5-6-7-     ';
      const unsub = '   ------------!     ';

      // This test is the same as the previous test, but the observable is
      // manipulated to make it look like an interop observable - an observable
      // from a foreign library. Interop subscribers are treated differently:
      // they are wrapped in a safe subscriber. This test ensures that
      // unsubscriptions are chained all the way to the interop subscriber.

      const result = e1.pipe(catchError(() => asInteropObservable(e2)));

      expectObservable(result, unsub).toBe(expected);
      expectSubscriptions(e1.subscriptions).toBe(e1subs);
      expectSubscriptions(e2.subscriptions).toBe(e2subs);
    });
  });

  it('should stop listening to a synchronous observable when unsubscribed', () => {
    const sideEffects: number[] = [];
    const synchronousObservable = concat(
      defer(() => {
        sideEffects.push(1);
        return of(1);
      }),
      defer(() => {
        sideEffects.push(2);
        return of(2);
      }),
      defer(() => {
        sideEffects.push(3);
        return of(3);
      })
    );

    throwError(new Error('Some error')).pipe(
      catchError(() => synchronousObservable),
      takeWhile((x) => x != 2) // unsubscribe at the second side-effect
    ).subscribe(() => { /* noop */ });

    expect(sideEffects).to.deep.equal([1, 2]);
  });

  it('should catch error and replace it with a hot Observable', () => {
    testScheduler.run(({ hot, expectObservable, expectSubscriptions }) => {
      const e1 = hot('  --a--b--#          ');
      const e1subs = '  ^-------!          ';
      const e2 = hot('  1-2-3-4-5-6-7-8-9-|');
      const e2subs = '  --------^---------!';
      const expected = '--a--b--5-6-7-8-9-|';

      const result = e1.pipe(catchError((err: any) => e2));

      expectObservable(result).toBe(expected);
      expectSubscriptions(e1.subscriptions).toBe(e1subs);
      expectSubscriptions(e2.subscriptions).toBe(e2subs);
    });
  });

  it('should catch and allow the cold observable to be repeated with the third ' +
  '(caught) argument', () => {
    testScheduler.run(({ cold, expectObservable, expectSubscriptions }) => {
      const e1 =  cold('--a--b--c--------|       ');
      const subs = [
        '               ^-------!                ',
        '              --------^-------!         ',
        '              ----------------^-------! '
      ];
      const expected = '--a--b----a--b----a--b--#';

      let retries = 0;
      const result = e1.pipe(
        map((n: any) => {
          if (n === 'c') {
            throw 'bad';
          }
          return n;
        }),
        catchError((err: any, caught: any) => {
          if (retries++ === 2) {
            throw 'done';
          }
          return caught;
        })
      );

      expectObservable(result).toBe(expected, undefined, 'done');
      expectSubscriptions(e1.subscriptions).toBe(subs);
    });
  });

  it('should catch and allow the hot observable to proceed with the third ' +
  '(caught) argument', () => {
    testScheduler.run(({ hot, expectObservable, expectSubscriptions }) => {
      const e1 = hot('  --a--b--c----d---|');
      const subs = [
        '               ^-------!         ',
        '              --------^--------! '];
      const expected = '--a--b-------d---|';

      let retries = 0;
      const result = e1.pipe(
        map((n: any) => {
          if (n === 'c') {
            throw 'bad';
          }
          return n;
        }),
        catchError((err: any, caught: any) => {
          if (retries++ === 2) {
            throw 'done';
          }
          return caught;
        })
      );

      expectObservable(result).toBe(expected, undefined, 'done');
      expectSubscriptions(e1.subscriptions).toBe(subs);
    });
  });

  it('should catch and replace a Observable.throw() as the source', () => {
    testScheduler.run(({ cold, expectObservable, expectSubscriptions }) => {
      const e1 = cold(' #');
      const subs = '    (^!)';
      const expected = '(abc|)';

      const result = e1.pipe(catchError((err: any) => of('a', 'b', 'c')));

      expectObservable(result).toBe(expected);
      expectSubscriptions(e1.subscriptions).toBe(subs);
    });
  });

  it('should mirror the source if it does not raise errors', () => {
    testScheduler.run(({ cold, expectObservable, expectSubscriptions }) => {
      const e1 = cold(' --a--b--c--|');
      const subs = '    ^----------!';
      const expected = '--a--b--c--|';

      const result = e1.pipe(catchError((err: any) => of('x', 'y', 'z')));

      expectObservable(result).toBe(expected);
      expectSubscriptions(e1.subscriptions).toBe(subs);
    });
  });

  it('should complete if you return Observable.empty()', () => {
    testScheduler.run(({ hot, cold, expectObservable, expectSubscriptions }) => {
      const e1 = hot('  --a--b--#');
      const e1subs = '  ^-------!';
      const e2 = cold('         |');
      const e2subs = '  --------(^!)';
      const expected = '--a--b--|';

      const result = e1.pipe(catchError(() => e2));

      expectObservable(result).toBe(expected);
      expectSubscriptions(e1.subscriptions).toBe(e1subs);
      expectSubscriptions(e2.subscriptions).toBe(e2subs);
    });
  });

  it('should raise error if you return Observable.throw()', () => {
    testScheduler.run(({ hot, cold, expectObservable, expectSubscriptions }) => {
      const e1 = hot('  --a--b--#');
      const e1subs = '  ^-------!';
      const e2 = cold('         #');
      const e2subs = '  --------(^!)';
      const expected = '--a--b--#';

      const result = e1.pipe(catchError(() => e2));

      expectObservable(result).toBe(expected);
      expectSubscriptions(e1.subscriptions).toBe(e1subs);
      expectSubscriptions(e2.subscriptions).toBe(e2subs);
    });
  });

  it('should never terminate if you return NEVER', () => {
    testScheduler.run(({ hot, cold, expectObservable, expectSubscriptions }) => {
      const e1 = hot('  --a--b--#');
      const e1subs = '  ^-------!';
      const e2 = cold('         -');
      const e2subs = '  --------^';
      const expected = '--a--b---';

      const result = e1.pipe(catchError(() => e2));

      expectObservable(result).toBe(expected);
      expectSubscriptions(e1.subscriptions).toBe(e1subs);
      expectSubscriptions(e2.subscriptions).toBe(e2subs);
    });
  });

  it('should pass the error as the first argument', (done: MochaDone) => {
    throwError('bad').pipe(
      catchError((err: any) => {
        expect(err).to.equal('bad');
        return EMPTY;
      })
    ).subscribe(() => {
    //noop
    }, (err: any) => {
      done(new Error('should not be called'));
    }, () => {
      done();
    });
  });

  it('should accept selector returns any ObservableInput', (done: MochaDone) => {
    const input$ = createObservableInputs(42);

    input$.pipe(
      mergeMap(input =>
        throwError('bad').pipe(catchError(err => input))
      )
    ).subscribe(x => {
      expect(x).to.be.equal(42);
    }, (err: any) => {
      done(new Error('should not be called'));
    }, () => {
      done();
    });
  });

  it('should catch errors throw from within the constructor', () => {
    // See https://github.com/zlq4863947/rxjs-cn//issues/3740
    testScheduler.run(({ expectObservable }) => {
      const source = concat(
        new Observable<string>(o => {
          o.next('a');
          throw 'kaboom';
        }).pipe(
          catchError(_ => of('b'))
        ),
        of('c')
      );
      const expected = '(abc|)';
      expectObservable(source).toBe(expected);
    });
  });

  context('fromPromise', () => {
    type SetTimeout = (callback: (...args: any[]) => void, ms: number, ...args: any[]) => NodeJS.Timer;

    let trueSetTimeout: SetTimeout;
    let sandbox: sinon.SinonSandbox;
    let timers: sinon.SinonFakeTimers;

    beforeEach(() => {
      trueSetTimeout = global.setTimeout;
      sandbox = sinon.sandbox.create();
      timers = sandbox.useFakeTimers();
    });

    afterEach(() => {
      sandbox.restore();
    });

    it('should chain a throw from a promise using Observable.throw', (done: MochaDone) => {
      const subscribeSpy = sinon.spy();
      const errorSpy = sinon.spy();
      const thrownError = new Error('BROKEN THROW');
      const testError = new Error('BROKEN PROMISE');
      from(Promise.reject(testError)).pipe(
        catchError(err =>
          throwError(thrownError)
        )
      ).subscribe(subscribeSpy, errorSpy);

      trueSetTimeout(() => {
        try {
          timers.tick(1);
        } catch (e) {
          return done(new Error('This should not have thrown an error'));
        }
        expect(subscribeSpy).not.to.be.called;
        expect(errorSpy).to.have.been.called;
        expect(errorSpy).to.have.been.calledWith(thrownError);
        done();
      }, 0);
    });
  });

  // TODO(v8): see https://github.com/zlq4863947/rxjs-cn//issues/5115
  // The re-implementation in version 8 should fix the problem in the
  // referenced issue. Closed subscribers should remain closed.
  /*
  it('issue #5115', (done: MochaDone) => {
    const source = new Observable<string>(observer => {
      observer.error(new Error('kaboom!'));
      observer.complete();
    });

    const sourceWithDelay = new Observable<string>(observer => {
      observer.next('delayed');
      observer.complete();
    }).pipe(delay(0));

    const values: string[] = [];
    source.pipe(
      catchError(err => sourceWithDelay)
    )
    .subscribe(
      value => values.push(value),
      err => done(err),
      () => {
        expect(values).to.deep.equal(['delayed']);
        done();
      }
    );
  });
  */
});


tools/prepare-package.js
const path = require('path');
const klawSync = require('klaw-sync');
const bo = require('@angular-devkit/build-optimizer');
const fs = require('fs-extra');
const { addLicenseTextToFile, addLicenseToFile } = require('./add-license-to-file');
const { cleanSourceMapRoot } = require('../.make-helpers');

const ROOT = 'dist/';
const CJS_ROOT = ROOT + 'cjs/';
const ESM5_ROOT = ROOT + 'esm5/';
const ESM2015_ROOT = ROOT + 'esm2015/';
const UMD_ROOT = ROOT + 'global/';
const ESM5_FOR_ROLLUP_ROOT = ROOT + 'esm5_for_rollup/';
const TYPE_ROOT = ROOT + 'types/';
const UMD_PKG = ROOT + 'bundles/';

// License info for minified files
let licenseUrl = 'https://github.com/zlq4863947/rxjs-cn//blob/master/LICENSE.txt';
let license = 'Apache License 2.0 ' + licenseUrl;

// Execute build optimizer transforms on ESM5 files
klawSync(ESM5_ROOT, {
  nodir: true,
  filter: function (item) {
    return item.path.endsWith('.js');
  }
})
  .map(item => item.path.slice((`${__dirname}/${ESM5_ROOT}`).length))
  .map(fileName => {
    if (!bo) return fileName;
    let fullPath = path.resolve(__dirname, ESM5_ROOT, fileName);
    // The file won't exist when running build_test as we don't create the ESM5 sources
    if (!fs.existsSync(fullPath)) return fileName;
    let content = fs.readFileSync(fullPath).toString();
    let transformed = bo.transformJavascript({
      content: content,
      getTransforms: [bo.getPrefixClassesTransformer, bo.getPrefixFunctionsTransformer, bo.getFoldFileTransformer]
    });
    fs.writeFileSync(fullPath, transformed.content);
    return fileName;
  });

if (fs.existsSync(UMD_ROOT)) {
  fs.copySync(UMD_ROOT, UMD_PKG);
  // Clean up source map paths so they can be re-mapped
  klawSync(UMD_PKG, { filter: (item) => item.path.endsWith('.js.map') })
    .map(f => f.path)
    .forEach(fName => {
      const sourceMap = fs.readJsonSync(fName);
      sourceMap.sources = sourceMap.sources.map(s => {
        const nm = 'node_modules/';
        const rr = path.resolve(ESM5_FOR_ROLLUP_ROOT);
        if (s.includes(nm)) {
          return s.substring(s.indexOf(nm) + nm.length);
        } else if (s.includes(rr)) {
          return s.substring(s.indexOf(rr) + rr.length);
        }
        return s;
      });
      fs.writeJsonSync(fName, sourceMap);
    });

  // Add licenses to tops of bundles
  addLicenseToFile('LICENSE.txt', UMD_PKG + 'rxjs.umd.js');
  addLicenseTextToFile(license, UMD_PKG + 'rxjs.umd.min.js');
  addLicenseToFile('LICENSE.txt', UMD_PKG + 'rxjs.umd.js');
  addLicenseTextToFile(license, UMD_PKG + 'rxjs.umd.min.js');
  cleanSourceMapRoot(UMD_PKG, CJS_ROOT);
}

// remove umd.js/umd.d.ts files that are only needed for creation of the umd bundle
fs.removeSync(CJS_ROOT + '/internal/umd.js');
fs.removeSync(CJS_ROOT + '/internal/umd.js.map');
fs.removeSync(ESM5_ROOT + '/internal/umd.js');
fs.removeSync(ESM5_ROOT + '/internal/umd.js.map');
fs.removeSync(ESM2015_ROOT + '/internal/umd.js');
fs.removeSync(ESM2015_ROOT + '/internal/umd.js.map');
fs.removeSync(TYPE_ROOT + '/internal/umd.d.ts');


src/internal/operators/mapTo.ts
import { Operator } from '../Operator';
import { Subscriber } from '../Subscriber';
import { Observable } from '../Observable';
import { OperatorFunction } from '../types';

/**
 * 每次源可观察对象发出值时，将值映射成常量。
 *
 * <span class="informal">类似 {@link map}, 但每次都会将每个源值映射为相同的输出值。</span>
 *
 * ![](mapTo.png)
 *
 * 以常量`值`作为参数，并在源可观测对象发出值时映射为此常量。也就是说，忽略实际的源值，只使用发射时机来发射指定的`值`。
 *
 * ## Example
 * 将每次点击都映射为字符串“Hi”
 * ```ts
 * import { fromEvent } from 'rxjs';
 * import { mapTo } from 'rxjs/operators';
 *
 * const clicks = fromEvent(document, 'click');
 * const greetings = clicks.pipe(mapTo('Hi'));
 * greetings.subscribe(x => console.log(x));
 * ```
 *
 * @see {@link map}
 *
 * @param {any} value 将每个源值映射成常量。
 * @return {Observable} 每次源可观察对象发出值时，都会映射成`指定的值`。
 * @method mapTo
 * @owner Observable
 */
export function mapTo<T, R>(value: R): OperatorFunction<T, R> {
  return (source: Observable<T>) => source.lift(new MapToOperator(value));
}

class MapToOperator<T, R> implements Operator<T, R> {

  value: R;

  constructor(value: R) {
    this.value = value;
  }

  call(subscriber: Subscriber<R>, source: any): any {
    return source.subscribe(new MapToSubscriber(subscriber, this.value));
  }
}

/**
 * We need this JSDoc comment for affecting ESDoc.
 * @ignore
 * @extends {Ignored}
 */
class MapToSubscriber<T, R> extends Subscriber<T> {

  value: R;

  constructor(destination: Subscriber<R>, value: R) {
    super(destination);
    this.value = value;
  }

  protected _next(x: T) {
    this.destination.next(this.value);
  }
}


src/internal/operators/filter.ts
import { Operator } from '../Operator';
import { Subscriber } from '../Subscriber';
import { Observable } from '../Observable';
import { OperatorFunction, MonoTypeOperatorFunction, TeardownLogic } from '../types';

/* tslint:disable:max-line-length */
export function filter<T, S extends T>(predicate: (value: T, index: number) => value is S,
                                       thisArg?: any): OperatorFunction<T, S>;
// NOTE(benlesh): T|null|undefined solves the issue discussed here: https://github.com/zlq4863947/rxjs-cn//issues/4959#issuecomment-520629091
export function filter<T>(predicate: BooleanConstructor): OperatorFunction<T|null|undefined, NonNullable<T>>;
export function filter<T>(predicate: (value: T, index: number) => boolean,
                          thisArg?: any): MonoTypeOperatorFunction<T>;
/* tslint:enable:max-line-length */

/**
 * Filter items emitted by the source Observable by only emitting those that
 * satisfy a specified predicate.
 *
 * <span class="informal">Like
 * [Array.prototype.filter()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter),
 * it only emits a value from the source if it passes a criterion function.</span>
 *
 * ![](filter.png)
 *
 * Similar to the well-known `Array.prototype.filter` method, this operator
 * takes values from the source Observable, passes them through a `predicate`
 * function and only emits those values that yielded `true`.
 *
 * ## Example
 * Emit only click events whose target was a DIV element
 * ```ts
 * import { fromEvent } from 'rxjs';
 * import { filter } from 'rxjs/operators';
 *
 * const clicks = fromEvent(document, 'click');
 * const clicksOnDivs = clicks.pipe(filter(ev => ev.target.tagName === 'DIV'));
 * clicksOnDivs.subscribe(x => console.log(x));
 * ```
 *
 * @see {@link distinct}
 * @see {@link distinctUntilChanged}
 * @see {@link distinctUntilKeyChanged}
 * @see {@link ignoreElements}
 * @see {@link partition}
 * @see {@link skip}
 *
 * @param predicate A function that
 * evaluates each value emitted by the source Observable. If it returns `true`,
 * the value is emitted, if `false` the value is not passed to the output
 * Observable. The `index` parameter is the number `i` for the i-th source
 * emission that has happened since the subscription, starting from the number
 * `0`.
 * @param thisArg An optional argument to determine the value of `this`
 * in the `predicate` function.
 */
export function filter<T>(predicate: (value: T, index: number) => boolean,
                          thisArg?: any): MonoTypeOperatorFunction<T> {
  return function filterOperatorFunction(source: Observable<T>): Observable<T> {
    return source.lift(new FilterOperator(predicate, thisArg));
  };
}

class FilterOperator<T> implements Operator<T, T> {
  constructor(private predicate: (value: T, index: number) => boolean,
              private thisArg?: any) {
  }

  call(subscriber: Subscriber<T>, source: any): TeardownLogic {
    return source.subscribe(new FilterSubscriber(subscriber, this.predicate, this.thisArg));
  }
}

/**
 * We need this JSDoc comment for affecting ESDoc.
 * @ignore
 * @extends {Ignored}
 */
class FilterSubscriber<T> extends Subscriber<T> {

  count: number = 0;

  constructor(destination: Subscriber<T>,
              private predicate: (value: T, index: number) => boolean,
              private thisArg: any) {
    super(destination);
  }

  // the try catch block below is left specifically for
  // optimization and perf reasons. a tryCatcher is not necessary here.
  protected _next(value: T) {
    let result: any;
    try {
      result = this.predicate.call(this.thisArg, value, this.count++);
    } catch (err) {
      this.destination.error(err);
      return;
    }
    if (result) {
      this.destination.next(value);
    }
  }
}
